Added Particle Emitter Pause option, and tests

PooperPig opened the pull request:
Added methods to Pause and UnPause emissions.
When paused, existing particles continue but new particles are not created.
UnPause continues to emit new particles.
Also adds IsPaused boolean.

Included Test/Demo of the effect in the tests.

I accidentally included changes to CCParticleSystemQuad (I made changes but then undid them - and did not intend to include them in the pull.

Commit:
Added Particle Emitter Pause option, and tests

minggo commented on cocos/2d/CCParticleSystem.h:
```diff
@@ -801,6 +801,17 @@ class CC_DLL ParticleSystem : public Node, public TextureProtocol, public Playab
+    /** Are the emissions paused
+     @return True if the emissions are paused, else false
+     */
+    virtual bool isPaused();
```
virtual bool isPaused() -> virtual bool isPaused() const?

minggo commented on cocos/2d/CCParticleSystem.h:
```diff
@@ -801,6 +801,17 @@ class CC_DLL ParticleSystem : public Node, public TextureProtocol, public Playab
+    virtual void pauseEmissions();
+    
+    /* UnPause the emissions*/
+    virtual void unPauseEmissions();
```
How about rename to `resumeEmissions()` ?

minggo commented on cocos/2d/CCParticleSystem.h:
```diff
@@ -956,6 +967,9 @@ class CC_DLL ParticleSystem : public Node, public TextureProtocol, public Playab
     PositionType _positionType;
+    
+    /** is the emitter paused */
+    bool paused;
```
paused -> _ paused to according cocos2d-x coding style

minggo commented on cocos/2d/CCParticleSystemQuad.cpp:
```diff
@@ -707,4 +707,10 @@ std::string ParticleSystemQuad::getDescription() const
+
+
+
+
```
It is better not add unneeded spaces.

minggo commented on cocos/2d/CCParticleSystemQuad.h:
```diff
@@ -144,6 +144,8 @@ class CC_DLL ParticleSystemQuad : public ParticleSystem
     virtual std::string getDescription() const override;
     
+
+    
```
Please don't add unneeded spaces.

Commit:
private field emitter => _emitter
isPaused changed to const
removed erroneously added spaces
renamed unPauseEmissions => resumeEmissions

minggo commented:
Thanks.

PooperPig commented:
No worries - thank you!

Max Young
mailmaxxx@gmail.com

> On 15 Jun 2016, at 8:31 PM, minggo notifications@github.com wrote:
> 
> Thanks.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub https://github.com/cocos2d/cocos2d-x/pull/15836#issuecomment-226149483, or mute the thread https://github.com/notifications/unsubscribe/ACTL4amQ01ZQAHbvcRQNg4iWhyzJMdNRks5qL9SegaJpZM4I1Lcq.

dumganhar commented on cocos/2d/CCParticleSystem.cpp:
```diff
@@ -588,6 +588,8 @@ ParticleSystem::~ParticleSystem()
 
 void ParticleSystem::addParticles(int count)
 {
+    if (_paused)
```
Why `_pause` wasn't initialized to false? It will be a random value (true / false).

dumganhar replied:
fix initialization problem at https://github.com/cocos2d/cocos2d-x/pull/15920
